### PR TITLE
Sound Optimizations and Debug Tools

### PR DESCRIPTION
**Pull request by timothymtorres**
## About The Pull Request
I was digging through sound code trying to add a few misc `.ogg`'s when I noticed a horrifying revelation. Sound calculations are broken as fuck. At first I didn't think it was a big deal until I remembered this comment in another PR:

chrome_6rspeYVig1

I asked for the profile logs:

```
                                                           Profile results (total time)
Proc Name                                                                                          Self CPU    Total CPU    Real Time     Overtime        Calls
----------------------------------------------------------------------------------------------    ---------    ---------    ---------    ---------    ---------
/proc/playsound                                                                                       4.771       26.841       26.902        0.347       153938
/mob/proc/playsound_local                                                                             9.511       15.689       15.891        0.724       429391

/mob/living/say                                                                                       0.125       19.064       19.945        0.000         2550
/mob/living/carbon/updatehealth                                                                       0.878        8.305        8.341        0.013        49847
/mob/living/carbon/proc/handle_organs                                                                 1.486        9.263        9.300        0.000        70977
/obj/item/organ/internal/on_life                                                                      0.415        0.455        0.561        0.000       340312
/mob/living/carbon/Life                                                                               0.758       30.085       30.123        0.000        71933
/mob/living/proc/Life                                                                                 5.509       39.404       39.512        0.000       401951
```

Yeah, that is _really_ bad. Especially since it is a hot proc that is used half a million times. Optimizing this would help a decent amount. So now you are probably wondering, what exactly is the problem?! Let's take a look.

https://www.desmos.com/calculator/sqdfl8ipgf

Sounds are being broadcast on a large radius which is typically about ~17 tiles. Since the volume of a sound is reduced by distance, the further away the sound is, the quieter it becomes. You would think, hey that's fine! Sounds become inaudible at the 17 tile distance limit right? Except no, they generally become inaudible about 2/3rds of the way there. (~10 tile range) Sound volume is between 1-100. When the sound volume is lower than 3, it becomes (pretty much) inaudible.

This means most of the sounds that are being spammed in game everywhere, you can't even hear!!! 

To fix this we used two major optimizations:

1. Add a `SOUND_AUDIBLE_VOLUME_MIN` to ignore any sound lower than this volume and calculate an audible distance
2. Use the new audible distance to drastically shorten the tile range of mob listeners using `get_hearers_in_view()`

To give you an idea of how much better this is:

Sounds that had a range of 17x17 (289 turfs) are now 12x12 (144 turfs). There is also large range machinery like the SM that has a 40x40 (1600 turfs) that is now 30x30 (900 turfs). 

It's quite an improvement.

I also went and added a debugging tool to the admin/debug equipment set that can be found in the debug box. It is a pair of earmuffs that when worn has any sounds sent to the mob via a `to_chat` message displaying the sounds max range, distance from player, volume, and sound name. It is highly recommend to walk while wearing the earmuffs since walking stops your mob from emitting footstep sounds.

## Why It's Good For The Game
The speed of sound is now faster than ever. Please don't break the sound barrier.

## Changelog
:cl:
refactor: Sound has been heavily optimized and will now ignore low volume sounds from far away.
admin: Add debugging sound earmuffs to admin equipment inside the debug box. Wear them to determine a sounds max range, distance, volume, and sound name. Highly recommended to walk otherwise you will get spammed with footstep sounds.
/:cl:

### CODE REVIEW

File: code/game/sound.dm [hunk 1]
```diff
@@ -151,9 +156,6 @@
 			sound_to_use.volume *= pressure_factor
 			//End Atmosphere affecting sound
 
-		if(sound_to_use.volume <= 0)
-			return //No sound
-
 		var/dx = turf_source.x - turf_loc.x // Hearing from the right/left
```
Comment (Kapu1178) [inline code/game/sound.dm:159]: You cannot remove this line. Pressure can reduce a sound volume to zero.

Comment (timothymtorres) [reply code/game/sound.dm]: I moved the volume check to a different scope to calculate audible range. I completely forgot about pressure decreasing volume.

Re-added via https://github.com/tgstation/tgstation/pull/88517/commits/0b42bdfb32d51c2a0408b4cee753e8dfecafb9e7

File: code/__DEFINES/sound.dm [hunk 2]
```diff
@@ -17,6 +17,44 @@
...
+ *
+ * Arguments:
+ * * volume: The initial volume of the sound being played
+ * * distance: How far away the sound is in tiles from the source
+ * * falloff_distance: Distance at which falloff begins. Sound is at peak volume (in regards to falloff) aslong as it is in this range.
+ * * falloff_exponent: Rate of falloff for the audio. Higher means quicker drop to low volume. Should generally be over 1 to indicate a quick dive to 0 rather than a slow dive.
+ * Returns: The max distance of a sound based on audible volume range
+ */
+#define CALCULATE_SOUND_VOLUME(volume, distance, max_distance, falloff_distance, falloff_exponent)\
+	((max(distance - falloff_distance, 0) ** (1 / falloff_exponent)) / ((max(max_distance, distance) - falloff_distance) ** (1 / falloff_exponent)) * volume)
```
Comment (LemonInTheDark) [inline code/__DEFINES/sound.dm:56]: why have these as macros? do you intend to reuse them elsewhere?

File: code/__DEFINES/sound.dm [hunk 3]
```diff
@@ -17,6 +17,44 @@
...
 
+/// This is the lowest volume that can be used by playsound otherwise it gets ignored
+/// Most sounds around 10 volume can barely be heard. Almost all sounds at 5 volume or below are inaudible
+/// This is to prevent sound being spammed at really low volumes due to distance calculations
+/// Recommend setting this to anywhere from 10-3 (or 0 to disable any sound minimum volume restrictions)
+/// Ex. For a 70 volume sound, 17 tile range, 3 exponent, 2 falloff_distance:
+/// Setting SOUND_AUDIBLE_VOLUME_MIN to 0 for the above will result in 17x17 radius (289 turfs)
+/// Setting SOUND_AUDIBLE_VOLUME_MIN to 5 for the above will result in 14x14 radius (196 turfs)
+/// Setting SOUND_AUDIBLE_VOLUME_MIN to 10 for the above will result in 11x11 radius (121 turfs)
+#define SOUND_AUDIBLE_VOLUME_MIN 3
```
Comment (LemonInTheDark) [inline code/__DEFINES/sound.dm:28]: bonus quest discussed on discord relayed for posterity, we need to normalize our sounds at some point

File: code/modules/clothing/ears/_ears.dm [hunk 4]
```diff
@@ -27,3 +27,8 @@
 	AddElement(/datum/element/earhealing)
 	AddComponent(/datum/component/wearertargeting/earprotection, list(ITEM_SLOT_EARS))
 	AddComponent(/datum/component/adjust_fishing_difficulty, -2)
+
+/obj/item/clothing/ears/earmuffs/debug
+	name = "debug earmuffs"
+	desc = "Wearing these sends a chat message for every sound played. Walking to ignore footsteps is highly recommended."
+	clothing_traits = list(TRAIT_SOUND_DEBUGGED)
```
Comment (LemonInTheDark) [inline code/modules/clothing/ears/_ears.dm:34]: v neat

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

**Comment by LemonInTheDark** [comment]
I will take this as a generalized good, if you want to profile you should make a proc that plays X sound Y times, and play it to a cube of mobs or something, it'll give you a best case for the optimization

**Comment by timothymtorres** [comment]
I went and profiled this using the full debug crew admin verb. At first it was showing no difference when I was profiling but this was due to the debug mobs that spawned not being inserted into a list used by the sound listeners. I fixed this via:
- #88694

This is the admin verb I used to spawn the sound 10,000 times.

```
ADMIN_VERB(cmd_admin_test_sound, R_DEBUG, "Test Sound", "Test Sound Nearby Mobs", ADMIN_CATEGORY_DEBUG, mob/living/carbon/human/H in world)
	message_admins(span_adminnotice("[key_name_admin(user)] tested the sound debug."))
	for(var/i in 1 to 10000)
		playsound(H, SFX_SEAR, 50, TRUE) // the default distance is 17 tiles
```

I also disabled client/can_hear checks on playsound_local so it would play the sound and not early return.
```
/mob/proc/playsound_local(turf/turf_source, soundin, vol as num, vary, frequency, falloff_exponent = SOUND_FALLOFF_EXPONENT, channel = 0, pressure_affected = TRUE, sound/sound_to_use, max_distance, falloff_distance = SOUND_DEFAULT_FALLOFF_DISTANCE, distance_multiplier = 1, use_reverb = TRUE)
	//if(!client || !can_hear())
	//	return
```

Here is my setup:

dreamseeker_SU9Gl1vM73

That is 10 crew spaced out with 1 tile gaps in between up to a distance of 20 tiles away.

---

Regular Sound
```
                                                           Profile results (total time)
Proc Name                                                                                          Self CPU    Total CPU    Real Time     Overtime        Calls
----------------------------------------------------------------------------------------------    ---------    ---------    ---------    ---------    ---------
/sound/New                                                                                            0.044        0.044        0.042        0.032        11821
/sound/proc/(init)                                                                                    0.026        0.027        0.026        0.019        11821
/proc/playsound                                                                                       0.822        4.947        4.949        0.691        11804
/datum/controller/subsystem/sounds/proc/random_available_channel                                      0.014        0.014        0.023        0.011        11765
/proc/get_sfx                                                                                         0.015        0.015        0.016        0.014        11109
/mob/proc/playsound_local                                                                             2.857        4.037        4.095        2.461       126637
/proc/get_rand_frequency                                                                              0.054        0.060        0.095        0.046       117969
```

---

Optimized Sound
```
                                                           Optimization Profile results (total time)
Proc Name                                                                                          Self CPU    Total CPU    Real Time     Overtime        Calls
----------------------------------------------------------------------------------------------    ---------    ---------    ---------    ---------    ---------
/sound/New                                                                                            0.037        0.039        0.032        0.034        11598
/sound/proc/(init)                                                                                    0.030        0.031        0.017        0.021        11598
/proc/playsound                                                                                       0.738        3.613        3.613        0.644        11584
/datum/controller/subsystem/sounds/proc/random_available_channel                                      0.009        0.009        0.030        0.007        11556
/proc/get_sfx                                                                                         0.014        0.015        0.014        0.014        11025
/mob/proc/playsound_local                                                                             2.088        2.796        2.860        1.895        78693
/proc/get_rand_frequency                                                                              0.029        0.034        0.057        0.029        71283

```

It's showing a performance of about 30%-50% faster for a lot of the procs. Keep in mind that this was tested playing a sound with `ignore_walls` which is the default argument for most sounds. If you were to set `ignore_walls = FALSE` the performance boost would be even larger since the sound would be using `get_hearers_in_view` instead of looping through the entire `clients_by_zlevel` list.

```
	if(ignore_walls)
		if(above_turf && istransparentturf(above_turf))
			listeners += SSmobs.clients_by_zlevel[above_turf.z]

		if(below_turf && istransparentturf(turf_source))
			listeners += SSmobs.clients_by_zlevel[below_turf.z]

	else //these sounds don't carry through walls
		listeners = get_hearers_in_view(audible_distance, turf_source)

		if(above_turf && istransparentturf(above_turf))
			listeners += get_hearers_in_view(audible_distance, above_turf)

		if(below_turf && istransparentturf(turf_source))
			listeners += get_hearers_in_view(audible_distance, below_turf)

	for(var/mob/listening_mob in listeners | SSmobs.dead_players_by_zlevel[source_z])//observers always hear through walls
		if(get_dist(listening_mob, turf_source) <= audible_distance)
			listening_mob.playsound_local(turf_source, soundin, vol, vary, frequency, falloff_exponent, channel, pressure_affected, S, maxdistance, falloff_distance, 1, use_reverb)
			. += listening_mob
```

File: code/__DEFINES/sound.dm (hunk 3 above)
Comment (timothymtorres) [reply code/__DEFINES/sound.dm:28]: I saw that but that looks like a job for an audio technician or sound engineer. You'd have to go over the hundreds of sounds in the codebase and modify each one.

File: code/__DEFINES/sound.dm (hunk 2 above)
Comment (timothymtorres) [reply code/__DEFINES/sound.dm:56]: The same reason as `CALCULATE_CONDUCTION_ENERGY` despite it being used in multiple spots. 

It's a clear reference to the algorithm used and supporting documentation. You don't want to stick all those code comments inside a proc.

**Pull request merged**